upgrade nacos client from 1.4.2 to 2.3.2

shalk opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->

<!-- ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👇 ====
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.
     ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👆 ==== -->

short story:   Nacos 2 is Better than Nacos 1.x. So update to Nacos 2.x is good choice.

long story:
 Nacos have released  2.x for long times,  the most important feature is  support gRPC for  "Long-Live" Connection.
https://www.alibabacloud.com/blog/an-in-depth-insight-into-nacos-2-0-architecture-and-new-model-with-a-supported-grpc-persistent-connection_597804

skywalking on nacos: 

Skywalking have some extension abilities with Nacos  Implemention. So Skywalking can use Nacos  for [Cluster Management](https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-cluster/#nacos) and [Dynamic Configuration](https://skywalking.apache.org/docs/main/next/en/setup/backend/dynamic-config/#dynamic-configuration-implementations)

After skywalking upgrade Nacos client ,  the performance  will be better when skywalking setup with nacos 2.x server




- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

wu-sheng commented on oap-server/server-cluster-plugin/cluster-nacos-plugin/pom.xml:
```diff
@@ -35,32 +35,6 @@
         <dependency>
             <groupId>com.alibaba.nacos</groupId>
             <artifactId>nacos-client</artifactId>
-            <exclusions>
-                <exclusion>
-                    <groupId>org.slf4j</groupId>
-                    <artifactId>slf4j-api</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>org.apache.logging.log4j</groupId>
-                    <artifactId>log4j-core</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>org.apache.logging.log4j</groupId>
-                    <artifactId>log4j-slf4j-impl</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>ch.qos.logback</groupId>
-                    <artifactId>logback-classic</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>ch.qos.logback</groupId>
-                    <artifactId>logback-core</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>net.jcip</groupId>
-                    <artifactId>jcip-annotations</artifactId>
-                </exclusion>
-            </exclusions>
```
Why do you exclude this?

shalk replied:
nacos-client 2.3.2 have make these package optional, not  transitivity

wu-sheng commented:
SkyWalking eyes will check the dependencies check. I believe you didn't do this.

https://github.com/apache/skywalking/blob/84dece92a65cbd727e1e9af66ef13a658f2f8978/.github/workflows/skywalking.yaml#L34-L44

wu-sheng commented:
Meanwhile, I don't see your explanation of this version upgrade. What is the impact of this upgrade?

wu-sheng commented:
Also, docs should be updated, https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-cluster/#nacos

Changes.md linked in the pull request has not been updated too.

wu-sheng commented on oap-server/server-cluster-plugin/cluster-nacos-plugin/src/test/java/org/apache/skywalking/oap/server/cluster/plugin/nacos/ClusterModuleNacosProviderFunctionalIT.java:
```diff
@@ -82,6 +81,13 @@ public void before() {
         Whitebox.setInternalState(telemetryModule, "loadedProvider", telemetryProvider);
         Mockito.when(moduleManager.find(TelemetryModule.NAME)).thenReturn(telemetryModule);
         nacosAddress = container.getHost() + ":" + container.getMappedPort(8848);
+        Integer nacosPortOffset = container.getMappedPort(9848) - container.getMappedPort(8848);
+        System.setProperty("nacos.server.grpc.port.offset", nacosPortOffset.toString());
```
What is this about?

shalk replied:
Nacos 2.x  server have two port, 8848(http) and 9948(gRPC) by default. Nacos Client 's implement use 8848(http.port) + 1000(nacos.server.grpc.port.offset) as gRPC port. It works on virtual machine.
 But the testContainer will not Mapport the port One-By-One. so i need to configure nacos.server.grpc.port.offset  

Doc ref :https://nacos.io/docs/next/manual/user/java-sdk/properties/#24-%E8%BF%9E%E6%8E%A5%E7%9B%B8%E5%85%B3

shalk commented:
Thank you for so quickly response, i will read these review and give some update later

wu-sheng commented:
Note  skywalking eyes tool is going to check the dependencies change, and License file needs to be updated.

shalk commented:
The check result of skywalking eyes  is as follow , it looks no license changes
```
➜  skywalking git:(update-nacos-to-2.3.2) docker run -it --rm -v $(pwd):/github/workspace ghcr.io/apache/skywalking-eyes/license-eye header check
INFO Loading configuration from file: .licenserc.yaml 
INFO Totally checked 3735 files, valid: 3223, invalid: 0, ignored: 512, fixed: 0 
➜  skywalking git:(update-nacos-to-2.3.2) docker run -it --rm -v $(pwd):/github/workspace ghcr.io/apache/skywalking-eyes/license-eye header fix  
INFO Loading configuration from file: .licenserc.yaml 
INFO Totally checked 3735 files, valid: 3223, invalid: 0, ignored: 512, fixed: 0 

```

wu-sheng commented:
This is header fix.(Source files have proper headers), But I mean this

https://github.com/apache/skywalking/blob/73555578d5b8ad05ae4eea189c3d94fd53171a26/.github/workflows/skywalking.yaml#L58-L87

shalk commented:
> This is header fix.(Source files have proper headers), But I mean this
> 
> https://github.com/apache/skywalking/blob/73555578d5b8ad05ae4eea189c3d94fd53171a26/.github/workflows/skywalking.yaml#L58-L87

Please check the license change. 

I am not sure what happend , so many nodejs related is deleted.

wu-sheng commented:
Have you ever build the project and run that? I doubt that.

Nodejs part is introduced from UI, which is a part of this project.

wu-sheng commented:
Could you explain why other libs out of nacos have been changed?

shalk commented:
These change may effect by nacos , I dig in deep.

1. commons-codec is chagned  to 1.15 . This is because nacos dep on commons-codes 1.5,  But module `apache-skywalking-apm` do not have the depedencyManagement. So the version in   oap-server-bom do not take effect.

```
INFO] ------------< org.apache.skywalking:apache-skywalking-apm >-------------
[INFO] Building apache-skywalking-apm 10.1.0-SNAPSHOT                   [89/89]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:3.6.0:resolve (default-cli) @ apache-skywalking-apm ---
[INFO] 
[INFO] The following files have been resolved:
[INFO]    org.apache.skywalking:server-starter:jar:10.1.0-SNAPSHOT:compile
...
[INFO]    com.alibaba.nacos:nacos-client:jar:2.3.2:compile
[INFO]    com.alibaba.nacos:nacos-auth-plugin:jar:2.3.2:compile
[INFO]    com.alibaba.nacos:nacos-encryption-plugin:jar:2.3.2:compile
[INFO]    commons-codec:commons-codec:jar:1.15:compile
```

2. commons-logging is deleted, because nacos-client do not dep on commons-logging any more.

3. org.apache.httpcomponents/httpasyncclient is not management in  any depedencyManagement, which is depedency by nacos-clien 

4. org.apache.httpcomponents/httpclient 4.5.13 , i think this is right version,  because ` <httpclient.version>4.5.13</httpclient.version>`

5. org.apache.httpcomponents/httpcore/4.4.16 ,  the problem is the same with commons-codec.


6. com.aayushatharva.brotli4j/native-osx-aarch64/1.15.0 because my local machine is macos, i will delete it  by manual



so for problem 
- 1,5 i will add  depedencyManagement in  module `apache-skywalking-apm` , 
- 3, i will add depency in bom,
- 2,4 is not problem
- 6 i will delete it .

wu-sheng commented on apm-dist/pom.xml:
```diff
@@ -34,6 +34,17 @@
             <activation>
                 <activeByDefault>true</activeByDefault>
             </activation>
+            <dependencyManagement>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.apache.skywalking</groupId>
+                        <artifactId>oap-server-bom</artifactId>
+                        <version>${project.version}</version>
+                        <scope>import</scope>
+                        <type>pom</type>
+                    </dependency>
+                </dependencies>
+            </dependencyManagement>
```
Why do we need to add this?

shalk replied:
module apache-skywalking-apm  without  dependencyManagement can not handle dependency conflict and version management

wu-sheng replied:
oap-server has that, and dist itself is just packaging, it should not include any new dependency AFAIK.

wu-sheng commented:
> com.aayushatharva.brotli4j/native-osx-aarch64/1.15.0 because my local machine is macos, this is not releated to nacos.

This should be removed. The CI check is running on Linux only. This dependency will trigger an error.

wu-sheng commented:
As expected, CI fails

```
diff --git a/dist-material/release-docs/LICENSE b/dist-material/release-docs/LICENSE
index 2e3a577..efce54e 100644
--- a/dist-material/release-docs/LICENSE
+++ b/dist-material/release-docs/LICENSE
@@ -214 +213,0 @@ The text of each license is the standard Apache 2.0 license.
-    https://mvnrepository.com/artifact/com.aayushatharva.brotli4j/native-osx-aarch64/1.15.0 Apache-2.0
```

wu-sheng commented:
About this PR still, you don't have 
1. Config and cluster management docs update about Nacos version requirements.
2. Change logs are not updated.
3. Please share your local relative tests about how this works on new Nacos server.

wu-sheng commented on docs/en/setup/backend/dynamic-config-nacos.md:
```diff
@@ -1,6 +1,10 @@
 # Dynamic Configuration Nacos Implementation
 
-[Nacos](https://github.com/alibaba/nacos) is also supported as a Dynamic Configuration Center (DCC). To use it, please configure it as follows:
+[Nacos](https://github.com/alibaba/nacos) is also supported as a Dynamic Configuration Center (DCC).
+
+**The Nacos client has upgraded to 2.x, so required Nacos Server version is  2.x**.
```
```suggestion
[Nacos](https://github.com/alibaba/nacos) 2.x is also supported as a Dynamic Configuration Center (DCC).
```

SkyWalking docs are versionized. So, this docs only show up when merged commit included version releases.

wu-sheng commented on docs/en/setup/backend/backend-cluster.md:
```diff
@@ -181,6 +181,8 @@ The following settings are provided to set the host and port manually, based on
 
 Set the **cluster/selector** to **nacos** in the yml to enable it.
 
+**The Nacos client has upgraded to 2.x, so required Nacos Server version is  2.x**.
```
```suggestion
Nacos 2.x is required.
```

wu-sheng commented:
This is still missing.

<img width="671" alt="image" src="https://github.com/apache/skywalking/assets/5441976/ae728fa9-7a69-4b84-b86f-954c74ff16a6">

shalk commented:
![1](https://github.com/apache/skywalking/assets/2435781/827b028c-6635-44e6-b234-a607a1e7bb4e)
![2](https://github.com/apache/skywalking/assets/2435781/ad6ced43-9844-4791-a8ce-6cecbeb10f51)

----

![skywalking-nacos-dynamic-config-1](https://github.com/apache/skywalking/assets/2435781/65dab25d-297c-4ad8-80c3-182e45d02414)
![skywalking-nacos-dynamic-config-2](https://github.com/apache/skywalking/assets/2435781/9b7b715b-67e6-4b71-b826-4e9ec24de376)

kezhenxu94 reviewed (commented):
Please also mention this is a breaking change in the CHANGES.md, users who are using Nacos now will be broken after upgrade

wu-sheng commented on docs/en/changes/changes.md:
```diff
@@ -29,5 +30,6 @@
 * Move the Official Dashboard docs to marketplace docs. 
 * Add marketplace introduction docs under `quick start` menu to reduce the confusion of finding feature docs.
 * Update Windows Metrics(Swap -> Virtual Memory)
+* [Break Change] Update Nacos version to 2.3.2.
```
Why UI matters about this?

wu-sheng commented on docs/en/changes/changes.md:
```diff
@@ -19,6 +19,7 @@
 * Add Python as a supported language for Pulsar.
 * Make more proper histogram buckets for the `persistence_timer_bulk_prepare_latency`,
   `persistence_timer_bulk_execute_latency` and `persistence_timer_bulk_all_latency` metrics in PersistenceTimer.
+* [Break Change] Update Nacos version to 2.3.2.
```
```suggestion
* [Break Change] Update Nacos version to 2.3.2. Nacod 1.x server can't serve as cluster coordinator and configuration server.
```

kezhenxu94 reviewed (approved):
Thanks!
